lib/archethic_web/controllers/api/origin_key_controller.ex

```diff
@@ -3,63 +3,82 @@ defmodule ArchethicWeb.API.OriginKeyController do
 
   alias Archethic.Crypto
   alias Archethic.SharedSecrets
-
-  alias Archethic.TransactionChain
-  alias Archethic.TransactionChain.TransactionData.Ownership
+  alias Archethic.TransactionChain.Transaction
+  alias Archethic.TransactionChain.TransactionData
+  alias ArchethicWeb.TransactionSubscriber
 
   def origin_key(conn, params) do
-    with %{"origin_public_key" => origin_public_key} <- params,
-         {:ok, origin_public_key} <- Base.decode16(origin_public_key, case: :mixed),
-         true <- Crypto.valid_public_key?(origin_public_key),
-         <<_curve_id::8, origin_id::8, _rest::binary>> <- origin_public_key,
-         {first_origin_family_public_key, _} <-
-           SharedSecrets.get_origin_family_from_origin_id(origin_id)
-           |> SharedSecrets.get_origin_family_seed()
-           |> Crypto.derive_keypair(0),
-         {:ok, tx} <-
-           Crypto.derive_address(first_origin_family_public_key)
-           |> TransactionChain.get_last_transaction(data: [:ownerships]),
-         ownership when ownership != nil <-
-           Enum.find(tx.data.ownerships, fn ownership ->
-             Ownership.authorized_public_key?(ownership, origin_public_key)
-           end) do
-      res = %{
-        encrypted_origin_private_keys: Base.encode16(ownership.secret),
-        encrypted_secret_key:
-          Ownership.get_encrypted_key(ownership, origin_public_key) |> Base.encode16()
-      }
+    {status_code, response} =
+      with %{"origin_public_key" => origin_public_key, "certificate" => certificate} <- params,
+           {:ok, origin_public_key} <- Base.decode16(origin_public_key, case: :mixed),
+           true <- Crypto.valid_public_key?(origin_public_key),
+           <<_curve_id::8, origin_id::8, _rest::binary>> <- origin_public_key do
+        origin_id
+        |> prepare_transaction(origin_public_key, certificate)
+        |> send_transaction()
+      else
+        error -> handle_error(error)
+      end
+
+    conn
+    |> put_status(status_code)
+    |> json(response)
+  end
+
+  defp prepare_transaction(origin_id, origin_public_key, certificate) do
+    signing_seed =
+      origin_id
+      |> Crypto.key_origin()
+      |> SharedSecrets.get_origin_family_seed()
+
+    {first_origin_family_public_key, _} = Crypto.derive_keypair(signing_seed, 0)
 
-      conn
-      |> put_status(:ok)
-      |> json(res)
+    address = Crypto.derive_address(first_origin_family_public_key)
+
+    with {:ok, last_address} <- Archethic.get_last_transaction_address(address),
```

Comment:
You don't need to get the last address of the chain as `Archethic.get_transaction_chain_length` always return the chain length whatever is the address.
Also I think that `origin` transactions will be stored by all node, so you may not need to ask chain index from other node but directly from the running node itself by using `TransactionChain.size(address)`

Reply:
Good point